File: src/MediaOrganizer.Shared/ViewModels/FailedFilesViewModel.cs
using Avalonia.Media.Imaging;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using MediaOrganizer.Core;
using MediaOrganizer.Core.Configuration;
using MediaOrganizer.Core.Execution;
using MediaOrganizer.Core.Logging;
using MediaOrganizer.Core.Models;
using MediaOrganizer.Core.Patterns;
using MediaOrganizer.Shared.Services;
using System.Collections.ObjectModel;

namespace MediaOrganizer.Shared.ViewModels;

/// <summary>失败文件列表项。IsChecked 为批量移动的目标选择（Android 触屏勾选；桌面单选不受影响）。</summary>
public partial class FailedItem : ObservableObject
{
    public FailedItem(UnparsedFile file)
    {
        File = file;
        Fingerprint = StructureFingerprint.Compute(file.File.FileName);
    }

    public UnparsedFile File { get; }
    public string Name => File.File.FileName;
    public string Path => File.File.Path;
    public long Size => File.File.Size;
    public string Fingerprint { get; }
    public string Reason => File.Reason;

    [ObservableProperty]
    private bool _isChecked;
}

/// <summary>
/// 失败文件：列表 + 预览 + 批量移动 + 打开系统应用（双端共享，ADR-0006 决策 4/5）。
/// 批量移动下沉 Core PendingFileMover（流式复制 + 删除源，SAF/本地统一语义）；
/// 确认弹窗 / 系统打开 / 缩略图预览经注入抽象（桌面实现齐全；Android 无预览、其余走系统 Intent）。
/// </summary>
public partial class FailedFilesViewModel : ViewModelBase
{
    private readonly AppConfig _config;
    private readonly AppLogger _logger;
    private readonly AppState _state;
    private readonly IConfirmDialog _confirm;
    private readonly ISystemFileOpener _opener;
    private readonly IImageLoader _imageLoader;

    public event Action<IEnumerable<string>>? NavigateToMagic;

    public ObservableCollection<FailedItem> Items { get; } = [];

    [ObservableProperty]
    private FailedItem? _selected;

    [ObservableProperty]
    private Bitmap? _preview;

    [ObservableProperty]
    private string _previewInfo = "选择文件查看预览";

    [ObservableProperty]
    private string _pendingDir;

    [ObservableProperty]
    private bool _isMoving;

    [ObservableProperty]
    private double _moveProgress;

    [ObservableProperty]
    private string _moveHint = "";

    public FailedFilesViewModel(
        AppState state,
        AppLogger logger,
        IConfirmDialog confirm,
        ISystemFileOpener opener,
        IImageLoader imageLoader)
    {
        _state = state;
        _config = state.Config;
        _logger = logger;
        _confirm = confirm;
        _opener = opener;
        _imageLoader = imageLoader;
        _pendingDir = state.Config.Paths.PendingDir;
    }

    public void Refresh(AnalysisResult? result)
    {
        Items.Clear();
        Preview = null;
        if (result is null) return;
        // FR-A6.1：按结构指纹聚类排序（指纹一致的相邻）
        foreach (var u in result.Unparsed
                     .OrderBy(f => StructureFingerprint.Compute(f.File.FileName))
                     .ThenBy(f => f.File.FileName))
            Items.Add(new FailedItem(u));
        _logger.Info($"失败文件列表已刷新：{Items.Count} 个（按指纹聚类排序）");
    }

    partial void OnSelectedChanged(FailedItem? value)
    {
        Preview = null;
        PreviewInfo = value is null ? "选择文件查看预览" : value.Path;
        if (value is null) return;
        _ = LoadPreviewAsync(value);
    }

    private async Task LoadPreviewAsync(FailedItem item)
    {
        var bmp = await Task.Run(() => _imageLoader.LoadThumbnail(item.Path, _config.General.PreviewSize));
        if (ReferenceEquals(Selected, item) && bmp is not null)
        {
            Preview = bmp;
            PreviewInfo = $"{item.Name}\n{item.Size:N0} 字节";
        }
    }

    [RelayCommand]
    private async Task OpenWithSystem(FailedItem? item = null)
    {
        var target = item ?? Selected;
        if (target is null) return;
        await _opener.OpenAsync(target.Path);
    }

    /// <summary>全选/取消全选（Android 触屏批量操作）：有未勾选项时全部勾选，否则全部取消。</summary>
    [RelayCommand]
    private void ToggleSelectAll()
    {
        var check = Items.Any(i => !i.IsChecked);
        foreach (var i in Items)
            i.IsChecked = check;
    }

    [RelayCommand]
    private void OpenMagicTools()
    {
        // 桌面端接线导航到魔术工具；Android 无魔术工具，不订阅此事件（FR-A6.4）
        var names = (Selected is not null
                ? new[] { Selected.Name }
                : Items.Select(i => i.Name))
            .Distinct()
            .Take(50)
            .ToArray();
        if (names.Length > 0)
            NavigateToMagic?.Invoke(names);
    }

    [RelayCommand]
    private async Task MoveToPending()
    {
        if (Items.Count == 0 || IsMoving) return;

        // 有勾选项时仅移动勾选的（Android 批量选择语义）；全部未勾选则移动全部（桌面语义保持不变）
        var items = Items.Where(i => i.IsChecked).ToArray();
        if (items.Length == 0) items = Items.ToArray();

        // FR-A6.5：目标目录必须由用户指定，不自动落盘兜底
        if (string.IsNullOrWhiteSpace(PendingDir))
        {
            MoveHint = "请先指定「待处理目录」";
            return;
        }
        var targetDir = PendingDir;

        // FR-A6.5：移动前确认（SAF 移动 = 复制 + 删除源）
        var confirmed = await _confirm.ConfirmAsync(
            "移入待处理目录",
            $"将把 {items.Length} 个文件移动到：\n{targetDir}\n\n（移动 = 复制 + 删除源）确定继续吗？");
        if (!confirmed) return;

        IsMoving = true;
        MoveProgress = 0;
        MoveHint = "";
        try
        {
            var target = StorageFactory.CreateLocalStorage(targetDir);
            var progress = new Progress<double>(p => MoveProgress = p * 100);
            var result = await new PendingFileMover()
                .MoveAsync(items.Select(i => i.File.File).ToArray(), target, progress);

            // 从列表移除已移动的条目
            foreach (var item in items.Where(i => result.MovedNames.Contains(i.Name)))
                Items.Remove(item);

            MoveHint = $"移动完成：成功 {result.Moved}，失败 {result.Failed}";
            _logger.Info(MoveHint);
            foreach (var err in result.Errors.Take(10))
                _logger.Warn(err);
        }
        catch (Exception ex)
        {
            MoveHint = $"移动失败：{ex.Message}";
            _logger.Error(MoveHint);
        }
        finally
        {
            IsMoving = false;
        }
    }
}

File: src-flutter/lib/viewmodels/failed_files_view_model.dart
/// 失败文件 VM（对应 Shared/ViewModels/FailedFilesViewModel.cs，FR-A6）：
/// 按结构指纹聚类排序的列表 + 勾选批量移动到待处理目录 + 系统应用打开。
library;

import 'dart:io';

import 'package:flutter/foundation.dart';

import '../core/app_state.dart';
import '../core/file_operator.dart';
import '../core/format.dart';
import '../core/models.dart';
import '../core/pattern_engine.dart' show structureFingerprint;
import '../core/pending_file_mover.dart';
import '../services/file_storage.dart';
import '../services/platform_services.dart';
import '../services/storage_factory.dart';

/// 失败文件列表项。isChecked 为批量移动的目标选择。
class FailedItem extends ChangeNotifier {
  final UnparsedFile file;
  final String fingerprint;
  bool isChecked = false;

  FailedItem(this.file) : fingerprint = structureFingerprint(file.file.fileName);

  String get name => file.file.fileName;
  String get path => file.file.path;
  int get size => file.file.size;
  String get reason => file.reason;
  String get sizeLabel => formatSize(size);

  void setChecked(bool v) {
    if (isChecked != v) {
      isChecked = v;
      notifyListeners();
    }
  }
}

class FailedFilesViewModel extends ChangeNotifier {
  final AppState state;

  /// 指纹 → 组内条目（保持指纹相邻排序后的分组视图）。
  final List<FailedItem> items = [];
  final expandedPaths = <String>{};

  bool isMoving = false;
  double moveProgress = 0; // 0~100
  String moveHint = '';
  CancelToken? _cancelToken;

  FailedFilesViewModel({required this.state});

  String get pendingDir => state.config.paths.pendingDir;

  int get checkedCount => items.where((i) => i.isChecked).length;

  void refresh(AnalysisResult? result) {
    items.clear();
    expandedPaths.clear();
    moveHint = '';
    if (result == null) {
      notifyListeners();
      return;
    }
    // FR-A6.1：按结构指纹聚类排序（指纹一致的相邻），组内按文件名
    final sorted = [...result.unparsed]
      ..sort((a, b) {
        final fa = structureFingerprint(a.file.fileName);
        final fb = structureFingerprint(b.file.fileName);
        final c = fa.compareTo(fb);
        return c != 0 ? c : a.file.fileName.compareTo(b.file.fileName);
      });
    for (final u in sorted) {
      items.add(FailedItem(u));
    }
    notifyListeners();
  }

  void toggleExpanded(String path) {
    if (!expandedPaths.add(path)) {
      expandedPaths.remove(path);
    }
    notifyListeners();
  }

  void toggleItem(FailedItem item) => item.setChecked(!item.isChecked);

  /// 全选/取消全选：有未勾选项时全部勾选，否则全部取消。
  bool get shouldSelectAll => items.any((i) => !i.isChecked);

  void toggleSelectAll() {
    final check = shouldSelectAll;
    for (final i in items) {
      i.setChecked(check);
    }
    notifyListeners();
  }

  /// 批量移动到待处理目录。确认弹窗由 UI 层完成后调用 [moveConfirmed]。
  List<FailedItem> itemsToMove() {
    final checked = items.where((i) => i.isChecked).toList();
    // 有勾选项时仅移动勾选的；全部未勾选则移动全部（桌面语义）
    return checked.isEmpty ? List.of(items) : checked;
  }

  Future<void> moveConfirmed(List<FailedItem> selected) async {
    if (items.isEmpty || isMoving || selected.isEmpty) return;

    final targetDir = pendingDir;
    if (targetDir.trim().isEmpty) {
      moveHint = '请先在工作台指定「待处理目录」';
      notifyListeners();
      return;
    }

    isMoving = true;
    moveProgress = 0;
    moveHint = '';
    _cancelToken = CancelToken();
    notifyListeners();
    try {
      final target = StorageFactory.createLocalStorage(targetDir);
      final result = await PendingFileMover().move(
        selected.map((i) => i.file.file).toList(),
        target,
        onProgress: (p) {
          moveProgress = p * 100;
          notifyListeners();
        },
        cancelToken: _cancelToken,
      );

      final movedNames = result.movedNames.toSet();
      items.removeWhere((i) => movedNames.contains(i.name));
      moveHint = '移动完成：成功 ${result.moved}，失败 ${result.failed}';
      if (result.errors.isNotEmpty) {
        moveHint += ' · ${result.errors.first}';
      }
    } catch (e) {
      moveHint = '移动失败：$e';
    } finally {
      isMoving = false;
      _cancelToken = null;
      notifyListeners();
    }
  }

  void cancelMove() => _cancelToken?.cancel();

  /// 用系统默认应用打开（FR-A6.3）。
  Future<void> openWithSystem(FailedItem item) async {
    if (!File(item.path).existsSync()) {
      moveHint = '文件不存在或不可访问：${item.path}';
      notifyListeners();
      return;
    }
    await ReportExporter.openWithSystem(item.path);
  }
}
